eyes_selenium/applitools/selenium/visual_grid/resource_cache.py

```diff
@@ -51,13 +51,24 @@ def __del__(self):
     def __iter__(self):
         return iter(self.cache_map)
 
-    def fetch_and_store(self, url, func_to_run, force=False):
-        if url in self and not force:
-            return self[url]
-        elif func_to_run:
-            self[url] = self.executor.submit(func_to_run, url)
+    def fetch_and_store(self, url, fetch_function, force=False):
+        # type: (typing.Text, typing.Callable[[typing.Text], VGResource], bool) -> bool
+        """
+        Schedules fetch of the url using fetch_function.
+
+        :param url: url to fetch resource from
+        :param fetch_function: function called to fetch the resource in a worker thread
+        :param force: option to force fetch even if the resource is
+                      already fetched or fetching
+
+        :return: True if the fetch was just scheduled or already in process
+                 False if it's already fetched and in the cache
+        """
+        with self.lock:
+            if not force and url in self.cache_map:
+                return isinstance(self.cache_map[url], Future)
+            self.cache_map[url] = self.executor.submit(fetch_function, url)
             return True
-        return False
```

Comment:
Nice docstring and solution!

In terms of types, `None` isn't a`False`. For example, mypy would suggest adding `-> Optional[bool]` to the types signature for the current code. I think it could be good practice to be more explicit in types even if it `None` which is count as `False` by default.

Reply:
But how on earth can it return None?

Reply:
Lol, looks like my eyes are twisted :D